src/design-system/components/chip/Chip.tsx

```diff
@@ -0,0 +1,233 @@
+import React, { CSSProperties, useState } from 'react'
+import classNames from 'classnames'
+import CircleIcon from '@mui/icons-material/Circle'
+import CloseRoundedIcon from '@mui/icons-material/CloseRounded'
+import ExpandMoreIcon from '@mui/icons-material/ExpandMore'
+
+import './Chip.scss'
+
+type ChipContentProps = {
+  label: string
+  startIcon?: React.ReactNode
+  endIcon?: React.ReactNode
+}
+
+type ChipType = 'filter' | 'input' | 'category' | 'state'
+
+type BaseChipProps = ChipContentProps & {
+  type: ChipType
+  size?: 'sm' | 'md' | 'lg'
+  disabled?: boolean
+}
+
+type FilterChipProps = BaseChipProps & {
+  type: 'filter'
+  options: string[]
+  variant?: 'filled' | 'minimal'
+}
+
+type InputChipProps = BaseChipProps & {
+  type: 'input'
+  variant?: 'filled' | 'outlined' | 'filled-outlined'
+}
+
+type ChipColor =
+  | 'blue-gray'
+  | 'turquoise'
+  | 'blue'
+  | 'green'
+  | 'yellow'
+  | 'purple'
+  | 'red'
+  | 'neutral'
+
+type CategoryChipProps = BaseChipProps & {
+  type: 'category'
+  detail: string
+  color?: ChipColor
+}
+
+type StateChipProps = BaseChipProps & {
+  type: 'state'
+  color?: ChipColor
+}
+
+export type ChipProps =
+  | FilterChipProps
+  | InputChipProps
+  | CategoryChipProps
+  | StateChipProps
+
+type CustomStyle = CSSProperties & {
+  [key: `--chip-${string}`]: string | undefined
+}
+
+const ChipContent: React.FC<ChipContentProps> = ({
+  label,
+  startIcon,
+  endIcon
+}) => (
+  <>
+    {startIcon && <span className='startIcon'>{startIcon}</span>}
+    <span className='label'>{label}</span>
+    {endIcon && <span className='endIcon'>{endIcon}</span>}
+  </>
+)
+
+const FilterChip: React.FC<FilterChipProps> = ({
+  label,
+  options = [],
+  variant = 'filled',
+  startIcon = <CircleIcon style={{ fontSize: 'inherit' }} />,
+  endIcon = <ExpandMoreIcon style={{ fontSize: 'inherit' }} />,
+  disabled = false,
+  size = 'md'
+}) => {
+  const [isOpen, setIsOpen] = useState(false)
+  const [selectedOption, setSelectedOption] = useState<string | null>(null)
+
+  const handleSelect = (option: string) => {
+    setSelectedOption(option)
+    setIsOpen(false)
+  }
+
+  const isSelected = !!selectedOption
+
+  const classes = classNames(
+    'chip',
+    `chip--${size}`,
+    `chip--filter`,
+    variant,
+    isSelected ? 'selected' : 'unselected',
+    {
+      disabled
+    }
+  )
+  return (
+    <div
+      className={classes}
+      onClick={() => !disabled && setIsOpen((prev) => !prev)}
+      tabIndex={0}
+    >
+      <ChipContent
+        endIcon={endIcon}
+        label={selectedOption || label}
+        startIcon={startIcon}
+      />
+      {isOpen && (
+        <ul className='dropdown-menu'>
+          {options.map((option) => (
+            <li
+              className='dropdown-item'
+              key={option}
+              onClick={() => {
```

Comment:
Same here - `useCallback`.
Always use `useCallback` when you're adding some new function inside React component. It is very useful for optimization.

Reply:
idk, in my opinion, using useCallback for this functions adds unnecessary complexity. The overhead of useCallback might not be worth it, cause the cost of the re-render is negligible.

Reply:
Oh, actually I used `useCallback` everywhere in my projects. I didn't notice that when you do not pass this callback function as a prop or as a dependency inside the dependency array, `useCallback` doesn't optimize this.
So thank you for reminding me of this)